airsonic-main/src/main/java/org/airsonic/player/controller/MainController.java

```diff
@@ -267,15 +290,63 @@ private String guessAlbum(List<MediaFile> children) {
-            List<MediaFile> sieblings = mediaFileService.getChildrenOf(parent, false, true, true);
-            result.addAll(sieblings.stream().filter(siebling -> siebling.isAlbum() && !siebling.equals(dir)).collect(Collectors.toList()));
+            List<MediaFile> siblings = mediaFileService.getChildrenOf(parent, false, true, true);
+            result.addAll(siblings.stream().filter(sibling -> sibling.isAlbum() && !sibling.equals(dir)).collect(Collectors.toList()));
+        }
+        return result;
+    }
+
+    private List<MediaFile> getAllAlbumDiscs(MediaFile dir) {
+        List<MediaFile> result = new ArrayList<>();
+        if (dir.getAlbumName() == null) {
+            result.add(dir);
+            return result;
+        }
+        if (dir.isAlbum() && dir.getDiscNumber() != null) {
+            MediaFile parent = mediaFileService.getParentOf(dir);
+            List<MediaFile> siblings = mediaFileService.getChildrenOf(parent, false, true, true);
+            result.addAll(siblings.stream().filter(sibling -> sibling.isAlbum() && dir.getAlbumName().equals(sibling.getAlbumName()) && sibling.getDiscNumber() != null).collect(Collectors.toList()));
         }
         return result;
     }
 
+    private void collapseMultiDiscAlbums(List<MediaFile> children) {
+        if (children != null && ! children.isEmpty()) {
+            // combine together discs of the same album
+            Map<String, List<MediaFile>> albumMap = new HashMap<>();
+            for (MediaFile child: children) {
+                if (child.isAlbum()&& child.getAlbumName() != null && child.getDiscNumber() != null && child.getDiscNumber() > 0) {
+                    List<MediaFile> albumList = albumMap.get(child.getAlbumName());
+                    if (albumList == null) {
+                        albumList = new ArrayList<MediaFile>();
+                        albumMap.put(child.getAlbumName(), albumList);
+                    }
+                    albumList.add(child);
+                }
+            }
+            // now go through and remove duplicates
+            for (List<MediaFile> albumList: albumMap.values()) {
+                if (albumList.size() > 1) {
+                    albumList.sort((a1,a2) -> a1.getDiscNumber() - a2.getDiscNumber());
+                    boolean keep = true;
+                    List<String> groupIdList = new ArrayList<>();
+                    for (MediaFile disc: albumList) {
+                        groupIdList.add(Integer.toString(disc.getId()));
+                        if (keep) {
+                            disc.setTitle(disc.getAlbumName() + " (" + albumList.size() + " discs)");
+                            disc.setAlbumDiscIds(groupIdList);
+                            keep = false;
+                        } else {
+                                children.remove(disc);
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This logic seems unnecessarily complex here. Basically, we are setting the first disc title to the disc album name with the number of discs and saving the accumulation of disc ids. That for sure can be done outside of the loop. Then we are discarding all other discs beside the first. I would break this into smaller chunks. Reiterate over the `albumList` if needed.

With that, I think this is the core logic of this PR. Basically when returning albums with discs, we collapse them all into one and on that one `MediaFile`, we set a list of ids of all media files for the disc. Seems like something we could do in the UI instead of making the returned data (both `siblingAlbums` and `subDirs` more confusing.

Reply:
This is in the Controller, so it really is in the UI already.  We could probably move it down to the Service/Model layer if we decided we wanted, say, the REST API or DLNA server to have access to it, but this seems like a web-UI-only type of feature.  And moving it up to the View would be a disaster; I mean, I could implement this in JSP or JavaScript, but it'd be like five times more confusing there.  At best.